Fix visibility of whitespace in config output

schneems opened the pull request:
When a config var has whitespace (especially trailing whitespace) it is hard to see. This commit wraps the values (if they exist) in a character so the difference is visually observable:

Before:

```
$ export PHP_INI_SCAN_DIR="/opt/homebrew/etc/php/8.4/conf.d         "
$ ./sapi/cli/php --ini
Configuration File (php.ini) Path: /usr/local/lib
Loaded Configuration File:         /opt/homebrew/etc/php/8.4/conf.d         
Scan for additional .ini files in: (none)
Additional .ini files parsed:      (none)
```

> [!NOTE]  
> The above output has trailing whitespace that is not visible, you can see it if you copy it into an editor.

After:

```
$ ./sapi/cli/php --ini
Configuration File (php.ini) Path: "/usr/local/lib"
Loaded Configuration File:         "/opt/homebrew/etc/php/8.4/conf.d         "
Scan for additional .ini files in: (none)
Additional .ini files parsed:      (none)
```

Above the whitespace is now visible `/opt/homebrew/etc/php/8.4/conf.d         `.

Close #18390

schneems commented on sapi/cli/php_cli.c:
```diff
@@ -1106,9 +1106,17 @@ static int do_cli(int argc, char **argv) /* {{{ */
+				} else {
+					zend_printf("Scan for additional .ini files in: (none)\n");
+				}
 				zend_printf("Additional .ini files parsed:      %s\n", php_ini_scanned_files ? php_ini_scanned_files : "(none)");
```
Mentioning a larger task that I intentionally did not take on in this PR in the interests in solving my initial issue as easily as possible: this can contain output that I did not wrap in backticks. 

```
$ export PHP_INI_SCAN_DIR="/opt/homebrew/etc/php/8.4/conf.d"
$ ./sapi/cli/php --ini
Configuration File (php.ini) Path: `/usr/local/lib`
Loaded Configuration File:         `/opt/homebrew/etc/php/8.4/conf.d`
Scan for additional .ini files in: (none)
Additional .ini files parsed:      /opt/homebrew/etc/php/8.4/conf.d/ext-opcache.ini
```

I feel that ideally, these Additional outputs (` /opt/homebrew/etc/php/8.4/conf.d/ext-opcache.ini`) would also be wrapped in backticks for consistency; however, implementing that would be more invasive. Also, since these are files that PHP found (as opposed to the config it was given), this wrapping is less helpful and more aesthetic.

TimWolla reviewed (commented):
Thank you for the PR. I think I would prefer double quotes as the indicator, as this would make the entire value copy and pasteable into a terminal without causing issues, i.e. I can just use the value and run `vim "whatever I copied"` to view a config file.

schneems commented:
@TimWolla seems good. I initially picked a backtick due to code literal format for markdown `like this` but I also agree that it's helpful to have a quote right there, especially for cases like `"/dirs/like this that have spaces/in them"` where copying and pasting the dir into a shell (a likely use case) would fail and the different segments might be treated as multiple arguments. 

I'll generally defer character choice as a style decision for primary project maintainers/contributors. Updated with a new commit and description to match.

Girgias reviewed (approved):
We usually use double quotes so this makes more sense than backticks.

Thank you!

nielsdos commented:
I wonder if the string inside the quotes should've gotten escaped

Girgias commented:
> I wonder if the string inside the quotes should've gotten escaped

Might be sensible, but isn't that a bit of a pre-existing issue?

nielsdos commented:
> > I wonder if the string inside the quotes should've gotten escaped
> 
> Might be sensible, but isn't that a bit of a pre-existing issue?

I agree it's pre-existing. I realized it just now because we now explicitly add quotes around it :slightly_smiling_face: I can look for making a patch if this sounds good.

Girgias commented:
> > > I wonder if the string inside the quotes should've gotten escaped
> > 
> > 
> > Might be sensible, but isn't that a bit of a pre-existing issue?
> 
> I agree it's pre-existing. I realized it just now because we now explicitly add quotes around it 🙂 I can look for making a patch if this sounds good.

Yeah it makes sense to me to escape it :)

schneems commented:
Thanks, all, for jumping on this quickly. Appreciate everything y'all do!

schneems commented:
> IMO this should be reverted, @nielsdos / @Girgias, until the BC impact has been thoroughly assessed.

Thankfully, this example works out of the box. It's a good thing @TimWolla advocated for quotes!

```
$ ./sapi/cli/php --ini
Configuration File (php.ini) Path: "/usr/local/lib"
Loaded Configuration File:         (none)
Scan for additional .ini files in: (none)
Additional .ini files parsed:      (none)
$ cut -d '"' -f 2 < <(./sapi/cli/php --ini | grep '(php.ini)' | sed -e "s|.*: s*||")

/usr/local/lib
```

I agree with @dzuelke that perhaps the impact area is larger than previously realized. I think, at minimum, adding a NEWS entry and mentioning that it is changing is worth doing. I'm happy to send a PR, but feel like I'm accidentally racking up commits for my mistake already. @dzuelke did a good job of finding that other bug, I would like him to have the commit. Here's a (perhaps overly verbose) suggestion:

```diff
$ git diff
diff --git a/NEWS b/NEWS
index 706533b9c8..c63b8d32c3 100644
--- a/NEWS
+++ b/NEWS
@@ -14,6 +14,10 @@ PHP                                                                        NEWS
   . Drop support for -z CLI/CGI flag. (nielsdos)
   . Fixed GH-17956 - development server 404 page does not adapt to mobiles.
     (pascalchevrel)
+  . Change `php --ini` output to now wrap values in double quotes to better
+    show hidden whitespace. Scripts that rely on parsing this information
+    can use the `cut` to remove the quotes. For example https://github.com/shivammathur/setup-php/blob/cf4cade2721270509d5b1c766ab3549210a39a2a/src/scripts/unix.sh#L246
+    (schneems)

 - CURL:
   . Added CURLFOLLOW_ALL, CURLFOLLOW_OBEYCODE and CURLFOLLOW_FIRSTONLY
```

Moving forward, this suggests that perhaps a more stable, machine-readable CLI flag (or command) for such script authors would be good for stability. For example, something like `php --ini=json` that authors could then parse via `jq` (which is a pretty common system dependency at this point). This could be a compliment to the `--ini=diff` just added in 8.5 (thanks @TimWolla, yet again).

dzuelke commented:
@schneems there is another option that I think solves both your original desire to highlight such a hard to catch error, and preserves backward compatibility: only quote the value if the TTY is a terminal.

That way, no regression is caused for any piping use case, but a human inspecting the output gets the quoting. This is similar to how many programs that output ANSI colors disable the once stdout is piped, so a perfectly common approach.

You can do it like so (stdout is always FD number 1):

```C
#ifdef PHP_WIN32
	/* Check if the Windows standard handle is redirected to file */
	stdout_is_tty = php_win32_console_fileno_is_console(1);
#elif HAVE_UNISTD_H
	/* Check if the file descriptor identifier is a terminal */
	stdout_is_tty = isatty(1);
#endif
```

And then use `php_escape_shell_arg()` on the string, it handles multibyte characters etc correctly and wraps everything in single quotes with correct escaping.

As I mentioned before, this quoting should maybe also be done for `php -i` (that code lives in `ext/standard/info.c`'s `php_print_info()`) since a lot of people use that as well, and there might be a few other "locations of interest" in there - `extension_dir`, `open_basedir`, `sys_temp_dir`, `upload_tmp_dir`, `user_dir` come to mind.
